// @eth-optimism/contracts-bedrock/dist/types/factories/contracts/dispute/interfaces/IBondManager__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type {
  IBondManager,
  IBondManagerInterface,
} from "../../../../contracts/dispute/interfaces/IBondManager";

const _abi = [
  {
    anonymous: false,
    inputs: [
      {
        indexed: false,
        internalType: "bytes32",
        name: "bondId",
        type: "bytes32",
      },
      {
        indexed: false,
        internalType: "address",
        name: "owner",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "expiration",
        type: "uint256",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "amount",
        type: "uint256",
      },
    ],
    name: "BondPosted",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: false,
        internalType: "bytes32",
        name: "bondId",
        type: "bytes32",
      },
      {
        indexed: false,
        internalType: "address",
        name: "claiment",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "amount",
        type: "uint256",
      },
    ],
    name: "BondReclaimed",
    type: "event",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: false,
        internalType: "bytes32",
        name: "bondId",
        type: "bytes32",
      },
      {
        indexed: false,
        internalType: "address",
        name: "owner",
        type: "address",
      },
      {
        indexed: false,
        internalType: "address",
        name: "seizer",
        type: "address",
      },
      {
        indexed: false,
        internalType: "uint256",
        name: "amount",
        type: "uint256",
      },
    ],
    name: "BondSeized",
    type: "event",
  },
  {
    inputs: [
      {
        internalType: "bytes32",
        name: "_bondId",
        type: "bytes32",
      },
      {
        internalType: "address",
        name: "_bondOwner",
        type: "address",
      },
      {
        internalType: "uint128",
        name: "_minClaimHold",
        type: "uint128",
      },
    ],
    name: "post",
    outputs: [],
    stateMutability: "payable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "bytes32",
        name: "_bondId",
        type: "bytes32",
      },
    ],
    name: "reclaim",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "bytes32",
        name: "_bondId",
        type: "bytes32",
      },
    ],
    name: "seize",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "bytes32",
        name: "_bondId",
        type: "bytes32",
      },
      {
        internalType: "address[]",
        name: "_claimRecipients",
        type: "address[]",
      },
    ],
    name: "seizeAndSplit",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
];

export class IBondManager__factory {
  static readonly abi = _abi;
  static createInterface(): IBondManagerInterface {
    return new utils.Interface(_abi) as IBondManagerInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): IBondManager {
    return new Contract(address, _abi, signerOrProvider) as IBondManager;
  }
}
